### PR TITLE
Use original map for data backing when possible.

### PR DESCRIPTION
**Pull request by jdalton**
This PR attempts to address #422 by using the original map for data backing when possible to avoid linear lookups.

### CODE REVIEW

**Comment by ljharb** [comment]
Thanks!

What about browsers where `-0` and `0` are not properly treated? There's also [insertion order](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L2681-L2693), browsers where iteration completion [throws a StopIterator exception](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L3341-L3343), browsers where "size" [is a function](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L3345) instead of a getter, browsers where Map and Set [aren't iterable](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L3348) except with an eval'd `for-of`, browsers that lack the facility to [clear and create iterators for keys/values/entries](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L3346) and then combined with the fact that the shim itself mutates `Map.prototype` when it already exists - so you'd need to cache all of the existing prototype methods (like `get`, `has`, `set`, etc), or else you'd risk an infinite loop.

In other words, using the original Map is a great approach, but there are a fractal of issues caused by the very necessity of the shim in the first place :-/

**Comment by jdalton** [comment]
> What about browsers where -0 and 0 are not properly treated?

I was hoping that was handled by the fast path. If not I'll rework to address it in the fast key path.

> There's also insertion order, browsers where iteration completion throws a StopIterator exception, 

Do you have a unit test to cover this? This is internal storage where the fast path has no concept of order since it's an empty object.

>  browsers where "size" is a function instead of a getter, browsers where Map and Set aren't iterable except with an eval'd for-of, 

I'm not using `size` of the internal map.

> browsers that lack the facility to clear and create iterators for keys/values/entries

I'm not sure that affects this internal use case.

> the shim itself mutates Map.prototype when it already exists - so you'd need to cache all of the existing prototype methods (like get, has, set, etc), or else you'd risk an infinite loop.

Ya, I'll address that.

**Comment by ljharb** [comment]
I have a number of test cases that should cover it - you'd need to run the test HTML in a very large number of browsers (which sadly travis doesn't cover) to check it out.

Mind you I'm not saying all these things will *definitely* cause a problem - more that it's possible, and I want to both be confident in this PR's correctness and also avoid leaving behind a footgun for the future.

**Comment by jdalton** [comment]
> Mind you I'm not saying all these things will definitely cause a problem - more that it's possible, and I want to both be confident in this PR's correctness and also avoid leaving behind a footgun for the future.

Cool. Ya, the use case I'm going for is to be a replacement for the linear search for object keys. That means it just needs to be a black box that is a yay or nay for if an object key is found. It can even go along with the array tracking as just a faster check but defer to the linear backing for when order is important.

**Comment by jdalton** [comment]
Looks like unit tests pass in IE11 and it does [resolve the perf issue](https://github.com/angular/angular/issues/8112#issuecomment-261681855) :tada:

**Comment by ljharb** [comment]
Thanks, I'll give it a thorough review, and do some extensive testing in as many browsers as I can before merging this.

**Comment by jdalton** [comment]
Tested on Safari 7, 8, 9, 10 and IE 9, 10 with no Map related fails.

**Comment by ljharb** [comment]
TC39 is this week so I'll need a bit more time to review; rest assured it's still on my list :-)

**Comment by ljharb** [comment]
Thanks!
